### ISSUE TITLE
Support GitLab flavored Markdown

### ISSUE THREAD

**Issue by engelgabriel** [RocketChat/Rocket.Chat#918]
https://github.com/gitlabhq/gitlabhq/blob/master/doc/markdown/markdown.md#special-gitlab-references

**Comment by geekgonecrazy** [comment]
@engelgabriel looks very similar to GitHub flavored markdown.  Do you know how it differs?

**Comment by engelgabriel** [comment]
I am not sure.. but we must improve our Markdown support in general.. add the tables, etc. There is probably a good package somewhere :)

**Comment by RichardFoxworthy** [comment]
@engelgabriel Thanks for great work on Rocket.Chat. Sorry if this a dumb question, I'm new to this... Should the existence of bounty appear automatically in GitHub issues? It does not look like it. This issue now has a bounty - https://www.bountysource.com/issues/26990858-support-full-markdown-spec-gitlab-github.

**Comment by engelgabriel** [comment]
Not sure why the plugin failed. I've edited it manually.

**Comment by scantem** [comment]
Tables would be very useful, I'm trying to write a better looking Redmine ticket announcement at the moment and being able to show the details in a table would look much better (and not use as many lines).

**Comment by RichardFoxworthy** [comment]
Mattermost has an excellent implementation, and I'd like to see equivalent for Rocket.Chat - see http://www.mattermost.org/open-source-slack-alternative-adopts-markdown/

Note that Mattermost has adopted Markdown for Channel header text as well as message text, which allows nice concise in-heading links.

Rocket.Chat could have equivalent treatment for the Topic text for each room

**Comment by engelgabriel** [comment]
I agree, let's use that as a reference.

**Comment by klatys** [comment]
If no one else is working on this, I'll give it a go

**Comment by geekgonecrazy** [comment]
:+1: I don't believe anyone is.  Feel free to take a crack at it! We're here if you need us.  Also on the demo server as usual :smile:

**Comment by klatys** [comment]
Ok. thanks guys. I started yesterday, but will be away till Sunday. I don't suppose fixing this next week will be a problem?

**Comment by engelgabriel** [comment]
Not a problem :)

**Comment by klatys** [comment]
Ok so I'm almost done with implementing it myself, then again, I started wondering whether or not it would be easier/cleaner to implement a library. [chjj/marked](https://github.com/chjj/marked) seems to be the one to do the trick... Is there some policy to this since all previous implementation has been done using custom code?

**Comment by LongLiveCHIEF** [comment]
@klatys the most popular implementation that I'm aware of is [codemirror](https://codemirror.net)

**Comment by klatys** [comment]
Hi, I commited markdown support, though against [mattermost's markdown](http://www.mattermost.org/open-source-slack-alternative-adopts-markdown/) (as mentioned above to be taken as a reference) support there are some issues/questions I'd like a second opinion to
- they use `four spaces as a code indent` which in my opinion colides with indentation of third level list so I would stick to using ticks only (as [standard GFM documentation does so](https://help.github.com/articles/basic-writing-and-formatting-syntax/))
- inlined image is always and not really inline and with full viewport width - ![inline image](http://image.prntscr.com/image/7b795317bc3b4928bf6aa5b5c08aa2df.png) is this intentional?

as for github issues / pull request:
- Since without pinging them i cannot know whether number is either issue or pull request i suggest differing them as gitlab does ![gitlab markdown for GH](http://image.prntscr.com/image/f0f81c15f4c54b93b16dda3423bb5fd5.png)
- also since rocket chat is not in a repository scope, there are two options to determine which repository to use
  - add default repository to settings
  - use reference containing repo - e.g. RocketChat/Rocket.Chat#918
- furthermore if noone is against I would add option to define whether to use github/bitbucket/?? repositories
- Gitlab allows Inline HTML... is that really a good idea?

**Comment by engelgabriel** [comment]
@RocketChat/core let's give some feedback here? This is important :)

**Comment by DeltaEvolution** [comment]
I have applied the commit of @klatys and i have the following error:
` Invalid regular expression: /(?:^|\s|\n)(#( **TEST**))[:.,s]?/: `
Nothing to repeat when I send # ** TEST ** in a channel

**Comment by LongLiveCHIEF** [comment]
> Regarding the linking of issues / commits. I kind of wonder if we shouldn't treat this as an integration? 

This is what we do now, and it works pretty well honestly.

**Comment by Sceat** [comment]
Any news ? the actual parsing is really weird :eyes:

**Comment by TheBelgarion** [comment]
i just stumbled on that issue because i tried converting jira-matter-bridge to work with rocket.chat and the drawback is the same without the markdown/table support it looks really ugly.

**Comment by LongLiveCHIEF** [comment]
I'm going to work on this over the weekend.  Plan is to implement codemirror insertion if code block is detected, and inject the code into the injected codemirror object.

**Comment by RichardFoxworthy** [comment]
Any news on this one?

**Comment by LongLiveCHIEF** [comment]
I had some issues setting up the dev environment the weekend after I posted this, and totally forgot about it.  I did get everything installed though, and nothing going on evening this week, so... yeah.

**Comment by LongLiveCHIEF** [comment]
ok... n/m... once again:

`curl: (7) Failed to connect to meteorinstall-4168.kxcdn.com port 443: Operation timed out`

I _hate_ meteor. _hate, hate, hate_.

**Comment by DeltaEvolution** [comment]
Any news @LongLiveCHIEF ?

**Comment by LongLiveCHIEF** [comment]
Nope.... still hating on meteor.

**Comment by LongLiveCHIEF** [comment]
Ok.. i just wasted *another* 3 hours trying to get meteor to *simply install on a machine*.  I've tried several Mac's (all post 2014 MBP's), and several Linux machines, including a few DO droplets just to make sure it's not something with my networking equipment.

They all get:

    curl: (7) Failed to connect to meteorinstall-4168.kxcdn.com port 443: Connection timed out
    Retrying download in 5 seconds...

    // repeated 9 more times

I have discovered this is pretty common with meteor installations, and it's not too often that these issues are resolved to the satisfaction of the user.
**l**
*then* I tried to create a new dev environment for this project by creating a [docker-compose extended override](https://docs.docker.com/compose/compose-file/), but using both [danieldent/meteor](https://hub.docker.com/r/danieldent/meteor/) and [meteorhacks/meteord](https://hub.docker.com/r/meteorhacks/meteord/).  In both cases here, mongo wouldn't start due to language issues.  I'm going to try one more thing... and if that doesn't work, I'll have spent more than an entire working day (collectively) just trying to _install a prerequisite_ for this project.

I really like Rocket.Chat... but this is my free time we're talking about, and I value it enough not to waste it carelessly.  If anyone has run into this issue while attempting to contribute to this project, I'd appreciate knowing what tactic you used to get around the transient failures of meteor's packaging cdn.

**Comment by geekgonecrazy** [comment]
@LongLiveCHIEF what OS are you downloading this on?  Never seen someone have this much trouble just getting meteor its self installed :(

**Comment by LongLiveCHIEF** [comment]
@geekgonecrazy nope. I have very fine control over my firewall, and I can see the traffic leaving just fine, and I am getting headers back from the CDN.  It's even keeping the initial connection open as it tries to re-attempt the download all 10 times.

I linked to an issue on the meteor/meteor site, and there are so many respondents that github collapses half of them in the middle and just displays a "134 more responses not shown" instead of making you scroll.

Many users seem to be reporting errors of failures on CI/CD builds as well... which are typically on a different network than their development machines.  I'm surprised you guys have never run into this.

I got it to install _once_... towards the end of July, but when I came back to this 2 weeks ago, and tried to update meteor, it bombed out both my installation and the download... and I haven't been able to install on any of my machines, either at home, or in the cloud, since.

**Comment by LongLiveCHIEF** [comment]
@geekgonecrazy 
See: https://github.com/LongLiveCHIEF/Rocket.Chat/commit/5687762dd9971eea370087c4a0a4404e26b2b8e0

I think I've had some luck by creating a docker-based environment to do development on.  Instead of installing the meteor binary, you can use a dockerized meteor distribution, and mount the Rocket.Chat app onto it.

I'm going to switch out the base of the build, and create a canonical `docker-compose.yml` and a `docker-compose.prod.yml`  if you like this idea and would like to see it go further.

**Comment by LongLiveCHIEF** [comment]
alright, so even my CI/CD servers can't download meteor, either from NY or San Francisco. I've now spent 4.5 hours on this, today alone.  I'm sorry, but I'm not going to try this again. Every.

I've seen this issue being reported by users as long as 2 years ago.  I'm not going to waste another moment trying.

Good luck Rocket.Chat devs/contributors! You are great, and I wish you the best.

**Comment by geekgonecrazy** [comment]
@LongLiveCHIEF regarding us not having the issue.  We have meteor cached in travis.  So i'm sure this helps keep us from hitting this as often. I think we've seen a few small occurrences in the past.  But usually an hour or so later it would be back.

@RocketChat/core as you can see a lot of things have been tried.  Ideas?

**Comment by LongLiveCHIEF** [comment]
Before I totally lost it, my next step was going to be to create a Docker image that builds meteor using the developer instructions, instead of the CDN install, and then modify my [`docker-compose.dev.yml`](https://github.com/LongLiveCHIEF/Rocket.Chat/blob/docker-compose-extend-environments/docker-compose.dev.yml) to use that as meteor.

**Comment by TheBelgarion** [comment]
had similar problems with meteor

using "54.192.225.217 warehouse.meteor.com" as resolve in /etc/hosts fixed that for me damm CDN still not working after month of having these problems

**Comment by LongLiveCHIEF** [comment]
thanks @TheBelgarion i'll give that a try tomorrow

**Comment by invious** [comment]
Is there table support? I can't get it to work if there is

**Comment by Flipez** [comment]
Any news on this one?

**Comment by alexweissman** [comment]
As a temporary workaround, you might consider [using LaTeX/KaTeX to generate tables](https://en.wikibooks.org/wiki/LaTeX/Tables).  Since Rocket _does_ support LaTeX, this is another way to generate tables.  Not as simple as with Markdown, of course...

**Comment by DeltaEvo** [comment]
@LongLiveCHIEF do you continue to work on that ?

**Comment by LongLiveCHIEF** [comment]
nope.  Too many issues just with getting meteor installed.  After wasting (literally) _days_ just trying to get the platform installed, I decided not to bother. I won't be contributing to any meteor-based project anytime soon... at least until the meteor team fixes their delivery issues that have been plaguing their community since 2014.

**Comment by nadavbrkt** [comment]
Are we getting anywhere with this? is there any reason not to edit [markdown.coffee](https://github.com/RocketChat/Rocket.Chat/blob/c0a3ca004fae1e118f60a59d035e56acdf9960cf/packages/rocketchat-markdown/markdown.coffee) to handle wanted markdown features?

**Comment by gromain** [comment]
I've been doing some tests with [markdown-it](https://github.com/markdown-it/markdown-it).
So far it's going well, but I still have some issues with styling.

However, I have some questions:
**What are the (for now) necessary features?**
I reckon at least as follows:
* bold
* italics
* strike
* inline code
* multiline code
* quote

**Should there be a complete compatibility with Github Flavored Markdown, or should we conform to CommonMark or a subset of both?**
For example, I'd like to have the checked lists fro GFM, but that's only a wish! ;-)

Also, an important question: **how are we going to treat the fact that the syntax is evolving?**
For example, in a lot of markdown parser I saw, bold is `**some bold text**` instead of `*some other bold text*` in Rocket.Chat right now. Will there be a change to be made in database (i.e. edit all messages to conform to new syntax where applicable) or will it just be "oh well, it will just look different"?

**Comment by gromain** [comment]
I'm fighting with the CSS now, where some settings (for ul and ol tags, for code tag) are overridden, but otherwise, it's working really well.

Is there some documentation somewhere regarding the organisation of CSS stylesheets in Rocket.Chat?

**Comment by gromain** [comment]
Hey everyone,

I have a preliminary version working well, the CSS may need a good bit of cleanup.
I'll put up a PR tomorrow I think.

Please, give your opinion regarding the questions in my previous comment:
- Should there be a complete compatibility with Github Flavored Markdown, or should we conform to CommonMark or a subset of both?
- How are we going to treat the fact that the syntax is evolving?

**Comment by gromain** [comment]
I created the PR related to this issue.
Please have a look and comment!

**Comment by TwizzyDizzy** [comment]
As discussed: please add a label "feature-needs-review".

/cc @SeanPackham 

Cheers
Thomas

**Comment by gromain** [comment]
Hi everyone,
After a break related to work, I'm planning on starting to work again on this.
Is there anyone already working on this? Anything I should take into account?

**Comment by broferek** [comment]
Nothing on my side but would really love to see this coming as a reality as the formatting with latex is awful when you want to try to have tables.

**Comment by subesokun** [comment]
@gromain Are there any updates from your side :)?

**Comment by fliptrail** [comment]
I think that we should use showdown js library.
https://github.com/showdownjs/showdown. It has support for almost every command and it is released under the MIT version. I am working to implement it.

**Comment by rodrigok** [comment]
I think we should have just one markdown implementation and have it well implemented in all clients.

**Comment by adrianovieira** [comment]
On iOS app ulist is working fine enough, but it doesn't work neither on browsers or desktop app.

iOS app working example:

```
* Item
  * Subitem
    * Subsubitem
* Next
```

**Comment by samtuke** [comment]
Indented lists do not work as documented in [RocketChat docs](https://docs.rocket.chat/contributors/contributing/documentation/markdown-styleguide#md005-no-inconsistent-indentation-for-list-items-at-the-same-level). 

Eg this:
![Selection_826](https://user-images.githubusercontent.com/695422/84685771-19dbfc00-af3b-11ea-89b4-3a963b69b0fa.png)

Is rendered as this (using RocketChat 2.3.2 Hosted):
![Selection_827](https://user-images.githubusercontent.com/695422/84685798-27918180-af3b-11ea-99a5-1651b018b3f6.png)

**Comment by sampaiodiego** [comment]
@samtuke that documentation is regarding the markdown used for our documentation pages, and not about the markdown support on the chat application. I'm sorry about the confusion

**Comment by samtuke** [comment]
@sampaiodiego Ah, me too then. I unthinkingly picked the first search engine result for 'RocketChat indented lists'

**Comment by adrianovieira** [comment]
> Indented lists do not work as documented in [RocketChat docs](https://docs.rocket.chat/contributors/contributing/documentation/markdown-styleguide#md005-no-inconsistent-indentation-for-list-items-at-the-same-level).
> 
> Eg this:
> ![Selection_826](https://user-images.githubusercontent.com/695422/84685771-19dbfc00-af3b-11ea-89b4-3a963b69b0fa.png)
> 
> Is rendered as this (using RocketChat 2.3.2 Hosted):
> ![Selection_827](https://user-images.githubusercontent.com/695422/84685798-27918180-af3b-11ea-99a5-1651b018b3f6.png)

Ok! 
this really is what we get on RcoketChat Desktop or web application.

![image](https://user-images.githubusercontent.com/5561794/84703104-0e360880-af2e-11ea-8957-67550b6c01b5.png)

But.. on the other hand, on mobile app we get this:

![image](https://user-images.githubusercontent.com/5561794/84703444-9e744d80-af2e-11ea-8911-384e9ca6df0c.png)

**Comment by saschawolff** [comment]
Please get the (un)ordered lists to work!
It would be really helpful to have them.

**Comment by dradux** [comment]
FWIW, I stumbled upon this issue yesterday trying to figure out how to make markdown work in the desktop app while it worked fine on the moble (android) app. In my case the issue was a setting in rocketchat desktop app:

administration > Message - scroll all the way down to the bottom to the Markdown section. I changed this from Original to Marked (you may want to check a few of the options there as well). After doing this my old messages which had markdown did not change but new messages did indent properly.

Before and after changing to Marked Markdown:
![2020-08-19_1134 43_600x439_scrot](https://user-images.githubusercontent.com/4999736/90656286-4f89b380-e210-11ea-8d38-a45c8109de7f.png)

**Comment by broferek** [comment]
Unfortunately, the workaround from @dradux doesn't work anymore in 5.0.4. No matter what options I checked, I can't get it to work properly.

Besides, there is a deprecation warning below the setting :
```This is a deprecated feature. It may not work as expected and will not get new updates```

This was the only way to get markdown in my messages and it was very useful  for displaying nice tables.

Anyone came up with another solution? It's really frustrating going back to plain text format when you were used to markdown.

**Comment by ulope** [comment]
It's really incomprehensible to me how such a core functionality can be left languishing for 7(!!!) years while ever more useless new "features" with more bugs are piled ontop.

**Comment by Nirajn2311** [comment]
Hi, any update on this and is anyone working on it? If not, I'd like to give it an attempt.

**Comment by saurabh47** [comment]
Any updates on feature implementation ? This issue is being open for more than 5 years. Any consideration in Product roadmap ?
Currently we are planning to setup chat server for our team. We are comparing between Rocketchat & mattermost. If this basic feature is not supported then it will be deal breaker.

**Comment by samtuke** [comment]
@saurabh47 Try Zulip

**Comment by pascalForGitHUB** [comment]
Any updates on this very important feature ?
